Issue warnings when accessing deprecated layer state

andy-sweet opened the pull request:
# References and relevant issues
See #5477 for some of the motivation here. This is split from that larger PR, to make reviewing the core idea here easier.

# Description
This introduces a mechanism to notify developers about deprecated state that is returned in `Layer.as_layer_data_tuple`. Similarly, this affects any plugins that rely on the `data, attrs, layer_type` tuple for layer interchange.

Without this mechanism, any equivalent parameters/properties that are deprecated will either be

1) present in the layer state dict returned, without a notification of their deprecation;
2) or removed from the layer state dict return.

We could live with (2), but it is a breaking change without the soft landing of a deprecation.

(1) is not much better because it just delays the removal and breaking change (unless we continue to accumulate all state ever in a layer, which seems unsustainable).

This PR also uses this new mechanism to ensure that the deprecated `Points.edge_*` state issues a warning when it is accessed.

andy-sweet committed:
Add deprecating dict

andy-sweet committed:
Use deprecating dict for points deprecation

andy-sweet committed:
More clean up

andy-sweet committed:
Add explanatory comment

andy-sweet commented on napari/utils/migrations.py:
```diff
@@ -211,3 +211,43 @@ def __init_subclass__(cls, **kwargs):
     return _OldClass
+
+
+class _DeprecatingDict(dict[str, Any]):
```
Unsure if this is the best way to sub-class a `dict` (e.g. should I extend a `MutableMapping` instead and implement that?). Open to other ideas.

Czaki replied:
if not dict then maybe `UserDict` https://docs.python.org/3/library/collections.html#userdict-objects

andy-sweet commented on napari/layers/_tests/test_serialize.py:
```diff
@@ -8,20 +8,32 @@
 
-@pytest.mark.parametrize(('Layer', 'data', 'ndim'), layer_test_data)
-def test_attrs_arrays(Layer, data, ndim):
+@pytest.mark.parametrize(('LayerType', 'data', 'ndim'), layer_test_data)
```
The name change here helps distinguish with the `Layer` type, which I introduced here to make the test a little more clear. This change does not need to occur in this PR.

andy-sweet commented on napari/utils/migrations.py:
```diff
@@ -211,3 +211,43 @@ def __init_subclass__(cls, **kwargs):
     return _OldClass
+
+
+class _DeprecatingDict(dict[str, Any]):
```
I really wanted to find an excuse to call this `SelfDeprecatingDict`, but that's not quite the right meaning here :p

andy-sweet committed:
Fix test

andy-sweet committed:
Remove typing on layer data tuple method

andy-sweet committed:
Fix some tests

andy-sweet committed:
Temporarily ignore failed test

Czaki commented on napari/utils/migrations.py:
```diff
@@ -211,3 +211,43 @@ def __init_subclass__(cls, **kwargs):
     return _OldClass
+
+
+class _DeprecatingDict(dict[str, Any]):
```
This will fail if someone pass it to layer constructor using `**kwargs` 


Iteration should return only non deprecated kwargs.

andy-sweet replied:
At first I did not like this suggestion because it changes the core behavior of `dict` (whereas the current approach only raises warnings under certain usage).

But I also understand that this is a critical issue. And the only time this object is created publicly is when calling `Layer.as_layer_data_tuple` or indirectly through plugins.

This approach may still causes issues for plugin authors that are iterating over the state `dict` and expecting the deprecated keys, but that's probably not many people.

Czaki replied:
Yes. And the current approach will create problems if someone does:

```python
data, _type, meta = Points.as_layer_data_tuple()

new_points = Points(data, **meta)
```
As it will end with exceptions. 


And this is exactly why `test_benchamrks` fails.

andy-sweet replied:
Understood. As previously mentioned and explained.

DragaDoncila reviewed (commented):
I like the general direction of this, I agree that since we're deprecating what's returned in the state dict, we should warn folks what's going to replace it. I think it goes hand in hand with the events, arguments, etc. - and thanks for giving this a bit of a shakedown!

I don't love how many spots in the layer file we now have to touch to ensure we get correct deprecation, but I'm not sure if there's a more concise way of doing it. Doesn't have to happen in this PR either, just flagging. 

What's holding this in draft mode @andy-sweet? I'm not sure I entirely understand the discussion re. `as_layer_data_tuple`, but were there any other concerns?

andy-sweet committed:
Store deprecated values separately

andy-sweet committed:
Add set, del, and contains

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Fix typing with cast

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Use FutureWarning instead of DeprecationWarning

andy-sweet committed:
Fix test

andy-sweet committed:
Use pytest warns instead of raises

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

Czaki commented on napari/utils/migrations.py:
```diff
@@ -218,3 +218,71 @@ def __init_subclass__(cls, **kwargs):
     return _OldClass
+
+
+class _DeprecatingDict(dict[str, Any]):
```
My main concern here is that part of our public API is class marked as a private one. Im not sure if it is valid approacg

andy-sweet replied:
Yeah, agreed. It's the one thing that stopped me from marking this as ready for review.

I think users of `as_layer_data_tuple` are pretty advanced, so I'm thinking about just dropping the `_` and making this public.

What do you think?

Czaki commented on napari/layers/surface/surface.py:
```diff
@@ -578,7 +579,7 @@ def _has_texture(self) -> bool:
         )
 
-    def _get_state(self) -> dict:
+    def _get_state(self) -> _DeprecatingDict:
```
maybe this PR should update this signature in all classes?

andy-sweet replied:
I can do that.

After some other changes. I might actually be able to downgrade the annotate type to `dict[str, Any]` which I'd prefer. Then those really interested in the implementation details can go find out what type it is really.

andy-sweet committed:
Update signature of get_state on all layers

andy-sweet committed:
Make DeprecatingDict public

andy-sweet committed:
Move and add tests

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

Czaki commented on napari/layers/base/base.py:
```diff
@@ -983,34 +984,39 @@ def _data_slice(self) -> _ThickNDSlice:
-            'projection_mode': self.projection_mode,
-        }
-        return base_dict
+        return DeprecatingDict(
```
I personally prefer to create here dict, and convert it to `DeprecatingDict` in `Points._get_state`.

andy-sweet replied:
Good idea. Especially since `Points` is the only class that uses this.

Czaki commented on napari/layers/points/points.py:
```diff
@@ -1474,6 +1475,15 @@ def _get_state(self):
+                    new_key=new_key,
+                    version='0.6.0',
+                    since_version='0.5.0',
+                )
```
This is the second place where this deprecation is defined. In https://github.com/andy-sweet/napari/pull/69 I have proposed a more generic solution to avoid multiple place to be edited.

andy-sweet commented on napari/utils/migrations.py:
```diff
@@ -218,3 +218,71 @@ def __init_subclass__(cls, **kwargs):
     return _OldClass
+
+
+class DeprecatingDict(dict[str, Any]):
```
I wonder if we should use `wrapt.ObjectProxy` instead of just overriding as the implementation is similar to other things we use that for: https://github.com/napari/napari/blob/f327d1f545b88c6473f3103cc46e8beac194d1ea/napari/utils/_proxies.py#L15

jni replied:
I don't have an opinion about it, but if this is working, we can just use it. It is nice and simple. Sometimes wrapt feels a little too magical 😅

Czaki committed:
Add more general mechanism for deprecating properies in state (#69)

# Description

Provide more generic mechanism to deprecate states, to have a single
source of truth

---------

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

andy-sweet commented on napari/utils/migrations.py:
```diff
@@ -137,6 +146,13 @@ def _setter(instance, value: Any) -> None:
+    if not hasattr(obj, '_deprecated_properties_list'):
+        obj._deprecated_properties_list = []
+
+    obj._deprecated_properties_list.append(
```
@Czaki : one problem with this approach is that it cannot handle deprecated properties (e.g. `current_edge_color` / `current_border_color`) that are not part of the state (i.e. returned by `get_state`).

I think that should be pretty easy with some modifications. Let me try it.

Czaki replied:
Why is it not part of state? It is deprecated using `add_deprecated_property` that updates this attribute.

andy-sweet replied:
I'm not certain why some of them are not state, but I think we would want the flexibility to have deprecated properties that are not state.

Czaki commented on napari/layers/base/base.py:
```diff
@@ -1020,6 +1025,16 @@ def _type_string(self) -> str:
+        if hasattr(self, '_deprecated_properties_list'):
+            state = DeprecatingDict(state)
+            element: DeprecatedProperty
+            for element in self._deprecated_properties_list:
```
```suggestion
            for element in self._deprecated_properties_list:
                if element.new_name not in state:
                    continue
```

@andy-sweet I think that it should solve the problem.

andy-sweet replied:
Yes, I have something similar.

Czaki commented on napari/layers/points/points.py:
It will be nice to have a test for at least one deprecated property if it is accesible in state dict.

andy-sweet replied:
Tests incoming!

andy-sweet commented on napari/layers/base/base.py:
```diff
@@ -1026,6 +1025,16 @@ def _type_string(self) -> str:
     def as_layer_data_tuple(self):
         state = self._get_state()
         state.pop('data', None)
+        if hasattr(self, '_deprecated_properties_list'):
```
After some more thought, I find that this approach makes things hard to understand (because the attr is only added/mutated in `Points._add_deprecated_properties`) and forces all layer types to understand it, so I'm going to push a new commit that uses `DeprecatedProperty`, but isolates some of the changes to `Points` and without so much repetition.

I think we will want a general purpose way to handling these across layers, but I don't think this is quite right yet. Open to alternative implementations that are easier to understand.

Happy to discuss more in a community meeting!

andy-sweet committed:
Refactor deprecated properties

Czaki commented on napari/layers/points/points.py:
```diff
@@ -68,6 +73,18 @@
+    'edge_contrast_limits',
+    'current_edge_color',
+    'edge_color_mode',
+)
```
This is unused

Czaki reviewed (commented):
I have found that the same problem as for Points layer, there is for `Image.interpolation`. So I propose a generalization based on `rename_argument` in https://github.com/andy-sweet/napari/pull/70

Czaki committed:
Merge branch 'main' into feat-layer-state-deprecation

jni committed:
Fix importerror and message for lxml_html_clean (#7017)

Closes #6798

See https://github.com/napari/napari/issues/6798#issuecomment-2181495221
for discussion.

psobolewskiPhD committed:
[UX/UI] flip the default for the reader plugin dialog to *not* have the Remember box checked (#7016)

# References and relevant issues
Closes: https://github.com/napari/napari/issues/7010

# Description
- flip the defaults for the reader plugin dialog for the `Remember this
plugin` checkbox from True (checked) to False (unchecked)
- flip the tests to match

psobolewskiPhD committed:
Make Shift-Up/Down in layer list expand and contract selection (#6606)

# References and relevant issues
Closes https://github.com/napari/napari/issues/6604

# Description
Qt behavior is for Shift-Up/Down to only expand selection. This isn't
consistent with file managers like macOS Finder or most (all?) text
editors, when selecting lines. E.g. Shift-Up adds next item to
selection, but then Shift-Down will remove it--and vice versa.
As a result in this PR I first prevent Qt from handling Shift-Up and
Shift-Down.
Then I tweak the (unused) `shift=True` behavior of layers.select_next
(and thus select_previous)
Finally I add Viewer keybindings to handle Shift-Up/Down to
select_next(shift=True) and select_previous(shift=True)
Note I hard-code these, so can't be user modified, like some of our
other base keybinds.

New behavior:


https://github.com/napari/napari/assets/76622105/aec778f7-42bc-4dd7-adf5-04780232f0ee

dalthviz committed:
Add tests for viewer scale bar attributes (`colored` and `ticks`) (#7006)

# References and relevant issues

Closes #6990 

# Description

Add test to check visibility of scale bar when changing viewer `colored`
and `ticks` attribute

Czaki committed:
Again fix translate of overlays (this time with tests) (#6927)

# References and relevant issues
closes #6897

fix again bug from #6633

# Description

This PR is fixing again problems with transforms of overlays, this time
it contains basic tests.

---------

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>
Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

Czaki committed:
Deprecated properies that handles also Image class (#70)

# References and relevant issues

https://github.com/napari/napari/pull/6976

# Description

I just realized that state is a set of arguments to the constructor, so
we should not depend on `add_deprecated_property` but on the results of
`rename_argument`

The `rename_argument` already defines `class.__init_._rename_argument`
attribute that is used for generation of `add_xxx` methods.

This approach of changes also handle renaming of `interpolation` to
`interpolation2d` in the `Image` class.

As an improvement, PR may contain an update of `rename_argument` to use
`DeprecatedProperty.message` in `warnings.warn` used in
`rename_argument`

andy-sweet committed:
Address PR feedback

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Revert unneeded changes

andy-sweet committed:
Rename property to parameter

Czaki commented on napari/layers/points/_tests/test_points.py:
```diff
@@ -2661,3 +2661,21 @@ def test_docstring():
+def test_as_layer_data_tuple_read_deprecated_attr(key: str):
+    layer = Points()
+    _, attrs, _ = layer.as_layer_data_tuple()
+    with pytest.warns(FutureWarning):
```
```suggestion
    with pytest.warns(FutureWarning, match="is deprecated since"):
```
Exact string may need to be tested, but it is much better to validate text of warning not to catch wrong warning.

Czaki commented on napari/layers/base/base.py:
```diff
@@ -1064,6 +1068,10 @@ def _type_string(self) -> str:
     def as_layer_data_tuple(self):
```
Maybe add type annotation?

andy-sweet replied:
I tried, but this triggers lots of typing issues unrelated to this PR, so it is best left for the future.

Czaki commented on napari/utils/_tests/test_migrations.py:
```diff
@@ -83,3 +85,49 @@ class macOS:
+    )
+    with pytest.warns(FutureWarning):
+        d['c'] = 3
+        assert d['c'] == 3
```
We should also test value of `d['a']` here, also `with pytest.warns(FutureWarning):` should handle only single line

jni replied:
@Czaki do you mean this?

```suggestion
    with pytest.warns(FutureWarning):
        d['c'] = 3
        assert d['c'] == 3
        assert d['c'] == d['a']
```

jni replied:
(and if so, feel free to commit the suggestion)

Czaki replied:
```suggestion
    with pytest.warns(FutureWarning):
        d['c'] = 3
    with pytest.warns(FutureWarning):
        assert d['c'] == 3
    assert d['a'] == 3
```
But the last line will not work with the current implementation. 

Having two lines under `with pytest.warns(FutureWarning):` pass if any of the lines emit warning.

andy-sweet replied:
Will update the tests to have separate `with pytest.warns`.

But I disagree about asserting `d['a'] == 3` because not all deprecations are renames / direct replacements.

So I don't think we should always try to enforce consistency between the value associated with the deprecated key and related non-deprecated one.

If the user sets a deprecated key, they will get a warning and we'll do our best not to break them. That seems like enough to me.

andy-sweet replied:
Another way of explaining my line of thinking here is that the current implementation of `DeprecatingDict` mostly just behaves like a `dict`. The exception is when you use a deprecated key, in which case it will treat those keys separately and will issue warnings. It will not change the behavior of `dict` when using non-deprecated keys under any circumstances.

By trying to enforce consistency between deprecated and non-deprecated keys, interacting with the deprecated keys has the potential side effect of mutating non-deprecated items. While this might be beneficial in some cases, I think thoses cases are limited in real world usage and I think the extra complication of deviating from `dict` is not worth it - especially since some deprecated keys may not have direct replacements.

Czaki commented on napari/utils/migrations.py:
```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
+
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self.deprecations = {}
```
```suggestion
        self._deprecations = {}
```
I think that this should be private value

jni replied:
It might actually be a useful interface for users to inspect things...

andy-sweet replied:
I'm a little undecided here.

I like this being public because it helps inspection/explanation and also gives simple flexibility on how things are deprecated.

But I like this being private because I don't want people adding arbitrary items to this dict.

Let me try making it private and adding some public helpers.

andy-sweet replied:
I made this private, added two separate public methods to add/set deprecated items, then added `deprecated_keys` as a public property. I think that works well enough.

Czaki commented on napari/utils/migrations.py:
```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
+        if key in self.deprecations:
+            value, message = self.deprecations[key]
+            warnings.warn(message, FutureWarning)
+            return value
```
```suggestion
            key, message = self._deprecations[key]
            warnings.warn(message, FutureWarning)
```
To enforce the same value for base and deprecated argument

andy-sweet replied:
In https://github.com/napari/napari/pull/5477 I'll need a mechanism to deprecate things that are not just renames, so this is not sufficient.

We could add a function to the deprecated value to transform the base/non-deprecated value, but then we'd also need a setter. In the future, even this behavior might be limited, so I'd want a more general way to store deprecated values that are not easily derivable from the base/non-deprecated values.

Czaki commented on napari/utils/migrations.py:
```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
+            _, message = self.deprecations[key]
+            warnings.warn(message, FutureWarning)
+            self.deprecations[key] = value, message
+            return None
```
```suggestion
            key, message = self._deprecations[key]
            warnings.warn(message, FutureWarning)
```

Czaki commented on napari/utils/migrations.py:
```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
+        return super().__delitem__(key)
+
+    def __contains__(self, key: object) -> bool:
+        if key in self.deprecations:
```
```suggestion
        if key in self._deprecations:
```

Czaki commented on napari/utils/migrations.py:
```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
+    def deprecate(self, param: DeprecatedParameter) -> None:
+        """Deprecates a key with a direct replacement using a corresponding message."""
+        self.deprecations[param.previous_name] = (
+            self[param.new_name],
```
```suggestion
            param.new_name,
```

Czaki commented on napari/utils/migrations.py:
```diff
@@ -218,3 +237,59 @@ def __init_subclass__(cls, **kwargs):
+            warnings.warn(message, FutureWarning)
+            del self.deprecations[key]
+            return None
+        return super().__delitem__(key)
```
Maybe `del` should also delete base key? 

I could imagine that deleting is to not pass given value to the layer constructor, and deleting deprecated properties without base will not be good.

andy-sweet replied:
I think it's fine as is. The user gets a warning telling them something is wrong and we try our best to support the deprecated behavior.

andy-sweet replied:
Unless we only support deprecations as renames, I think it will be tricky to implement something that will work well generally here. Another option is to raise with the deprecation message, which will break plugins that do this instead of warning them and continuing without effect (as now).

Czaki replied:
As I understand, our only usecase is rename of parameter?

andy-sweet replied:
There are other use cases in #5477

Czaki reviewed (commented):
Looks good, but I prefer to stronger connect deprecated property with new name in `DeprecatingDict` as for my intuition, current construction may lead to unpredicted behaviour. 

I also prefer to `pytest.warns(FutureWarning):` to be improved to contain `match` argument for better validation of warning.

jni committed:
add match for deprecation message testing

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jni commented on napari/components/_tests/test_layers_list.py:
```diff
@@ -427,7 +427,10 @@ def test_layers_save_selected(builtins, tmpdir, layer_data_and_types):
-@pytest.mark.filterwarnings('ignore:`np.int` is a deprecated alias for')
+# TODO: this warning filter can be removed when a new version
+# of napari-svg includes the following PR:
+# https://github.com/napari/napari-svg/pull/38
```
napari-svg 0.2.0 is out, but we need to update the test requirements to remove this. Maybe worth leaving until a follow-up PR.

jni reviewed (approved):
I'm happy with this as is, but I'll leave things until morning in case @andy-sweet wants to take one or more of @Czaki's suggestions.

andy-sweet committed:
Use separate pytest.warns

andy-sweet committed:
Address PR feedback and simplify

andy-sweet committed:
Merge branch 'feat-layer-state-deprecation' of github.com:andy-sweet/napari into feat-layer-state-deprecation

andy-sweet committed:
Use match in other tests

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation

andy-sweet committed:
Bring back named tuple for renamed attribute

andy-sweet committed:
Only handle renames

Czaki commented on napari/layers/points/_tests/test_points.py:
```diff
@@ -2661,3 +2661,21 @@ def test_docstring():
+)
+def test_as_layer_data_tuple_read_deprecated_attr(key: str):
+    layer = Points()
+    _, attrs, _ = layer.as_layer_data_tuple()
```
```suggestion
    _, attrs, _ = Points().as_layer_data_tuple()
```

Czaki commented on napari/layers/surface/surface.py:
```diff
@@ -592,12 +592,12 @@ def _has_texture(self) -> bool:
         Returns
         -------
-        state : dict
+        state : dict[str, Any]
```
Looking at numpydoc reference, it should look like 
```suggestion
        state : dict of str to Any
```

https://numpydoc.readthedocs.io/en/latest/format.html#parameters

If you agree, please update all type annotations

andy-sweet replied:
Seems fine. Will do.

Czaki commented on napari/utils/migrations.py:
```diff
@@ -218,3 +245,71 @@ def __init_subclass__(cls, **kwargs):
     return _OldClass
+
+
+class DeprecatingDict(dict[str, Any]):
```
I'm not sure if this should be part of a public API. 
At least I will put in docstring, that is not intended to be used outside napari as may be mutated. 

But it is not a strong opinion.

andy-sweet replied:
I'm fine adding a leading `_` and will add to the docstring to clarify that the user should use this as as if it were a `dict[str, Any]`.

andy-sweet replied:
An instance of this will be publicly accessible though.

Czaki replied:
Yes. But I prefer an instance of `_DeprecatingDict` being publicly available rather than people importing it and using it in different contexts.

andy-sweet committed:
Refactor implementation

andy-sweet committed:
Apply suggestions from code review

Co-authored-by: Grzegorz Bokota <bokota+github@gmail.com>

andy-sweet committed:
Merge branch 'feat-layer-state-deprecation' of github.com:andy-sweet/napari into feat-layer-state-deprecation

andy-sweet committed:
Update docstring return types

andy-sweet committed:
Privatize deprecating dict

andy-sweet commented on napari/utils/migrations.py:
```diff
@@ -218,3 +245,71 @@ def __init_subclass__(cls, **kwargs):
     return _OldClass
+
+
+class DeprecatingDict(dict[str, Any]):
```
Sub-classing `dict` means that we don't get consistent implementations of `pop`, `get`, etc. for free as those don't necessarily call our definitions of `__setitem__` and `__getitem__`.

Instead, we'll need to sub-class `UserDict` or `MutableMapping`, which seems preferable over manually overriding `dict` methods.

This means that the return type of `_get_state` should be changed from `dict[str, Any]` to `MutableMapping[str, Any]` or `UserDict[str, Any]`. That seems fine to me, but wanted to confirm.

andy-sweet replied:
I'd personally prefer `MutableMapping[str, Any]` since it has a smaller API, but that could break people that have been relying on some `dict` specific methods (e.g. `copy`)

andy-sweet committed:
Implement with UserDict to get pop, get, etc.

andy-sweet committed:
Only rename key if in renamed

andy-sweet committed:
Fix names of tests

andy-sweet committed:
Merge branch 'main' into feat-layer-state-deprecation
